sw/airborne/modules/gpio_ext/gpio_ext_common.c

```diff
@@ -0,0 +1,196 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+/** @file "modules/gpio_ext/gpio_ext_common.c"
+ * @author Tom van Dijk <tomvand@users.noreply.github.com>
+ * Common external GPIO functions.
+ */
+
+#include "gpio_ext_common.h"
+
+#include "generated/airframe.h"
+#include "mcu_periph/gpio.h"
+#include "led.h"
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <assert.h>
+
+
+/* External GPIO providers */
+#include "gpio_ext_pca95xx.h"
+
+#define GPIO_EXT_NOT_PROVIDED NULL
+#define GPIO_EXT_PCA95XX &pca95xx_functions
+
+
+#ifndef GPIO_EXT_PROVIDER1
+#define GPIO_EXT_PROVIDER1 GPIO_EXT_NOT_PROVIDED
+#endif
+#ifndef GPIO_EXT_PROVIDER2
+#define GPIO_EXT_PROVIDER2 GPIO_EXT_NOT_PROVIDED
+#endif
+#ifndef GPIO_EXT_PROVIDER3
+#define GPIO_EXT_PROVIDER3 GPIO_EXT_NOT_PROVIDED
+#endif
+#ifndef GPIO_EXT_PROVIDER4
+#define GPIO_EXT_PROVIDER4 GPIO_EXT_NOT_PROVIDED
+#endif
+
+static struct gpio_ext_functions *gpio_ext_impl[GPIOEXT_NB] = {
+    GPIO_EXT_PROVIDER1,
+    GPIO_EXT_PROVIDER2,
+    GPIO_EXT_PROVIDER3,
+    GPIO_EXT_PROVIDER4,
+};
+
+
+/* External GPIO configuration */
+// Not all defines need to be used by all implementations
+#ifndef GPIO_EXT_BLOCKING1
+#define GPIO_EXT_BLOCKING1 TRUE
+#endif
+#ifndef GPIO_EXT_BLOCKING2
+#define GPIO_EXT_BLOCKING2 TRUE
+#endif
+#ifndef GPIO_EXT_BLOCKING3
+#define GPIO_EXT_BLOCKING3 TRUE
+#endif
+#ifndef GPIO_EXT_BLOCKING4
+#define GPIO_EXT_BLOCKING4 TRUE
+#endif
+const bool gpio_ext_blocking[] = {
+    GPIO_EXT_BLOCKING1,
+    GPIO_EXT_BLOCKING1,
+    GPIO_EXT_BLOCKING1,
+    GPIO_EXT_BLOCKING1,
+};
+
+#ifndef GPIO_EXT_I2C_PERIPH1
+#define GPIO_EXT_I2C_PERIPH1 NULL
+#endif
+#ifndef GPIO_EXT_I2C_PERIPH2
+#define GPIO_EXT_I2C_PERIPH2 NULL
+#endif
+#ifndef GPIO_EXT_I2C_PERIPH3
+#define GPIO_EXT_I2C_PERIPH3 NULL
+#endif
+#ifndef GPIO_EXT_I2C_PERIPH4
+#define GPIO_EXT_I2C_PERIPH4 NULL
+#endif
+struct i2c_periph *gpio_ext_i2c_periph[] = {
+    GPIO_EXT_I2C_PERIPH1,
+    GPIO_EXT_I2C_PERIPH2,
+    GPIO_EXT_I2C_PERIPH3,
+    GPIO_EXT_I2C_PERIPH4,
+};
+
+#ifndef GPIO_EXT_I2C_ADDRESS1
+#define GPIO_EXT_I2C_ADDRESS1 0x00
+#endif
+#ifndef GPIO_EXT_I2C_ADDRESS2
+#define GPIO_EXT_I2C_ADDRESS2 0x00
+#endif
+#ifndef GPIO_EXT_I2C_ADDRESS3
+#define GPIO_EXT_I2C_ADDRESS3 0x00
+#endif
+#ifndef GPIO_EXT_I2C_ADDRESS4
+#define GPIO_EXT_I2C_ADDRESS4 0x00
+#endif
+const uint8_t gpio_ext_i2c_addr[] = {
+    GPIO_EXT_I2C_ADDRESS1,
+    GPIO_EXT_I2C_ADDRESS2,
+    GPIO_EXT_I2C_ADDRESS3,
+    GPIO_EXT_I2C_ADDRESS4,
+};
+
+
+
+/*
+ * External GPIO implementation
+ */
+#define SAFE_CALL(_port, _fn, _args...) if(gpio_ext_impl[_port - GPIOEXT1] && gpio_ext_impl[_port - GPIOEXT1]->_fn) gpio_ext_impl[_port - GPIOEXT1]->_fn(_args);
+
+// Wrapping functions
+void __wrap_gpio_setup_output(uint32_t port, uint32_t gpios);
```

Comment:
The original gpio functions have a `gpio_port_t` type for the port (but not always...) which is arch dependent. Did you test this with the regular stm32 arch, chibios or both ?

Reply:
I know, I saw that the function arguments are different between architectures, but was hesitant to change anything because it might break other people's code. The current uint32_t does not match the stm32 functions either iirc, but still seems to work and is large enough for all architectures. I have tested this only on the stm32 arch.

Reply:
then I don't think this will work with chibios since we can overload functions

Reply:
I have never used chibios before so I can't really comment, but looking at the code I think that the only problem right now is that the gpio_arch  functions are static? Apart from that I think it should work, it would just redirect the gpio calls to the existing arch functions unless things are being written to a GPIOEXT port.

What do you mean by overloading functions? It looks like normal c code to me? (arch/chibios/mcu_periph/gpio_arch.c/h)

Reply:
the ports in chibios are `ioport_t` which are pointers to structures, incompatible with uint32, that's why we have this `gpio_port_t` typedef at some places
but I wouldn't bother to much since it seems that latest chibios versions are not doing differently

Reply:
Ok thanks, then I understand. I'll leave it like this for now.